[SPARK-6890] [core] Fix launcher lib work with SPARK_PREPEND_CLASSES.

vanzin opened the pull request:
The fix for SPARK-6406 broke the case where sub-processes are launched
when SPARK_PREPEND_CLASSES is set, because the code now would only add
the launcher's build directory to the sub-process's classpath instead
of the complete assembly.

This patch fixes the problem by having the launch scripts stash the
assembly's location in an environment variable. This is not the prettiest
solution, but it avoids having to plumb that location all the way through
the Worker code that launches executors. The env variable is always
set by the launch scripts, so users cannot override it.

Commit:
[SPARK-6890] [core] Fix launcher lib work with SPARK_PREPEND_CLASSES.

The fix for SPARK-6406 broke the case where sub-processes are launched
when SPARK_PREPEND_CLASSES is set, because the code now would only add
the launcher's build directory to the sub-process's classpath instead
of the complete assembly.

This patch fixes the problem by having the launch scripts stash the
assembly's location in an environment variable. This is not the prettiest
solution, but it avoids having to plumb that location all the way through
the Worker code that launches executors. The env variable is always
set by the launch scripts, so users cannot override it.

vanzin commented:
I tested on Linux with and without SPARK_PREPEND_CLASSES. I'll try it on Windows tomorrow just to make sure I didn't break anything there.

/cc @davies @andrewor14

SparkQA commented:
[Test build #30214 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30214/consoleFull) for   PR 5504 at commit [`31d3ce8`](https://github.com/apache/spark/commit/31d3ce857c3e8b481706770a88efc222a6655f5d).

SparkQA commented:
[Test build #30214 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30214/consoleFull) for   PR 5504 at commit [`31d3ce8`](https://github.com/apache/spark/commit/31d3ce857c3e8b481706770a88efc222a6655f5d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class MinOf(left: Expression, right: Expression) extends Expression `

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30214/
Test FAILed.

Commit:
Merge branch 'master' into SPARK-6890

vanzin commented:
Update: tested on Windows too, looks fine.

SparkQA commented:
[Test build #30258 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30258/consoleFull) for   PR 5504 at commit [`ff87a60`](https://github.com/apache/spark/commit/ff87a60c2d00ab9e00a44d5f656f905401f47d69).

SparkQA commented:
[Test build #30258 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30258/consoleFull) for   PR 5504 at commit [`ff87a60`](https://github.com/apache/spark/commit/ff87a60c2d00ab9e00a44d5f656f905401f47d69).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30258/
Test FAILed.

andrewor14 commented:
Test output was weird. Jenkins retest this please

SparkQA commented:
[Test build #30271 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30271/consoleFull) for   PR 5504 at commit [`ff87a60`](https://github.com/apache/spark/commit/ff87a60c2d00ab9e00a44d5f656f905401f47d69).

andrewor14 commented:
The changes LGTM. I'm testing this out locally.

SparkQA commented:
[Test build #30271 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30271/consoleFull) for   PR 5504 at commit [`ff87a60`](https://github.com/apache/spark/commit/ff87a60c2d00ab9e00a44d5f656f905401f47d69).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30271/
Test FAILed.

andrewor14 commented:
This works locally. I wonder why it's failing random MLlib tests.

andrewor14 commented:
retest this please

SparkQA commented:
[Test build #30273 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30273/consoleFull) for   PR 5504 at commit [`ff87a60`](https://github.com/apache/spark/commit/ff87a60c2d00ab9e00a44d5f656f905401f47d69).

SparkQA commented:
[Test build #30273 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30273/consoleFull) for   PR 5504 at commit [`ff87a60`](https://github.com/apache/spark/commit/ff87a60c2d00ab9e00a44d5f656f905401f47d69).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30273/
Test FAILed.

andrewor14 commented:
The test results are really flaky, and I don't believe these tests are failing outside of this PR. For this reason I don't think we should merge this patch as is yet, but I haven't dug into why the MLlib tests are failing.

vanzin commented:
Let me see if I can reproduce it locally.

vanzin commented:
All tests fail with an NPE in lines that are calling `SparkEnv.get`... shouldn't be caused by this change, but well, I'll take a look anyway.

vanzin commented:
Actually, it might be caused by this change.

Commit:
Fix tests.

Ignore fact that assembly location may not be in the environment when
running tests. Also handle the case where user code may end up calling
this code path, by restoring the code that looks up the spark assembly
under SPARK_HOME.

vanzin commented:
The latest change should fix the tests. I realize that now there's some duplication, and that because I had to restore the Java code to look for the assembly (see comment in patch), we don't necessarily need to expose it in an env variable. But being able to control the assembly's location from the shell scripts was the whole point of #5085, so I guess if we want to support that use case, the env variable is still needed.

SparkQA commented:
[Test build #30281 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30281/consoleFull) for   PR 5504 at commit [`7aec921`](https://github.com/apache/spark/commit/7aec9216d770d4e7b5ea8e2b67741f0e06ddf723).

SparkQA commented:
[Test build #30281 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30281/consoleFull) for   PR 5504 at commit [`7aec921`](https://github.com/apache/spark/commit/7aec9216d770d4e7b5ea8e2b67741f0e06ddf723).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch **adds the following new dependencies:**
   * `snappy-java-1.1.1.7.jar`

 * This patch **removes the following dependencies:**
   * `snappy-java-1.1.1.6.jar`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30281/
Test PASSed.

andrewor14 commented:
Alright, I tested the latest changes locally again and can verify that this patch does fix the problem. Now that it's also passing tests I will merge this into master. Thanks for fixing this so promptly @vanzin.

nishkamravi2 commented:
@vanzin I'm ok with the introduction of an spark assembly env var but it does seem unnecessary. The logic could have been as simple as: 
if (SPARK_PREPEND_CLASSES) 
   find_assembly() 
else 
   getLocation().getPath()

nishkamravi2 commented on launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java:
```diff
@@ -186,12 +186,24 @@ void addOptionString(List<String> cmd, String options) {
+    String assembly = getenv(ENV_SPARK_ASSEMBLY);
+    if (assembly == null && isEmpty(getenv("SPARK_TESTING"))) {
+      assembly = findAssembly();
+    }
```
Why do we need  "&& isEmpty(getenv("SPARK_TESTING")"  ?

vanzin replied:
This has already been pushed, but it's explained in the big comment right before the code.

nishkamravi2 replied:
if (assembly == null) findAssembly() ?

vanzin replied:
This code has already been pushed.

andrewor14 replied:
we can always push a hot fix

vanzin replied:
But there is nothing to fix.

We don't want to look for the assembly when tests are running because it may not exist. Tests do a lot of "new SparkContext()" with master = `local-cluster[blah]`, and this check ensures those tests work even if the assembly is not there.

nishkamravi2 replied:
I don't think this check 'ensures' that those tests work, this check 'requires' that those tests work if the assembly is not there (more like an assert). I don't feel strongly for or against it, but it does seem unnecessary.

nishkamravi2 commented:
Will submit a hot fix in a bit

srowen commented:
I think @vanzin is saying the logic is correct? at least the intent is to not bother searching for the assembly when testing. Do you mean you think that has to change?

nishkamravi2 commented:
Maintaining a non-null value for spark_assembly is a good thing unless there is a good reason to not do so.

vanzin commented:
The reason is explained in the code comments and in my comments above. Please don't fix what doesn't need fixing.
